### ISSUE TITLE
OpenSSL v1.1.0 compatibility issues

### ISSUE THREAD

**Issue by zwoop** [IBM-Swift/BlueSSLService#38]
I'm getting the following errors, on Fedora26, which uses OpenSSL v1.1.0:

```
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:318:54: error: use of undeclared type 'SSL'
                public private(set) var cSSL: UnsafeMutablePointer<SSL>? = nil
                                                                   ^~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:322:49: error: use of undeclared type 'SSL_METHOD'
                public private(set) var method: UnsafePointer<SSL_METHOD>? = nil
                                                              ^~~~~~~~~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:325:57: error: use of undeclared type 'SSL_CTX'
                public private(set) var context: UnsafeMutablePointer<SSL_CTX>? = nil
                                                                      ^~~~~~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:1120:80: error: use of undeclared type 'SSL'
        private func prepareConnection(socket: Socket) throws -> UnsafeMutablePointer<SSL> {
                                                                                      ^~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:405:5: error: use of unresolved identifier 'SSL_library_init'
                                SSL_library_init()
                                ^~~~~~~~~~~~~~~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:406:5: error: use of unresolved identifier 'SSL_load_error_strings'
                                SSL_load_error_strings()
                                ^~~~~~~~~~~~~~~~~~~~~~
OpenSSL.ERR_lib_error_string:1:13: note: did you mean 'ERR_lib_error_string'?
public func ERR_lib_error_string(_ e: UInt) -> UnsafePointer<Int8>!
            ^
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:407:5: warning: 'OPENSSL_config' is deprecated
                                OPENSSL_config(nil)
                                ^
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:408:5: error: use of unresolved identifier 'OPENSSL_add_all_algorithms_conf'
                                OPENSSL_add_all_algorithms_conf()
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:415:19: error: use of unresolved identifier 'SSLv23_server_method'
                                self.method = SSLv23_server_method()
                                              ^~~~~~~~~~~~~~~~~~~~
OpenSSL.SSLv3_server_method:2:13: note: did you mean 'SSLv3_server_method'?
public func SSLv3_server_method() -> OpaquePointer!
            ^
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:419:19: error: use of unresolved identifier 'SSLv23_client_method'
                                self.method = SSLv23_client_method()
                                              ^~~~~~~~~~~~~~~~~~~~
OpenSSL.SSLv3_client_method:2:13: note: did you mean 'SSLv3_client_method'?
public func SSLv3_client_method() -> OpaquePointer!
            ^
/homer/leif/swift/Kitura/.build/checkouts/BlueSSLService.git--774207894858193348/Sources/SSLService.swift:853:27: error: use of unresolved identifier 'SSL_CTRL_OPTIONS'
                                SSL_CTX_ctrl(context, SSL_CTRL_OPTIONS, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_COMPRESSION, nil)
                                                      ^~~~~~~~~~~~~~~~
OpenSSL.SSL_CTRL_CHAIN:1:12: note: did you mean 'SSL_CTRL_CHAIN'?
public var SSL_CTRL_CHAIN: Int32 { get }
           ^
OpenSSL.SSL_CTRL_SET_MTU:1:12: note: did you mean 'SSL_CTRL_SET_MTU'?
public var SSL_CTRL_SET_MTU: Int32 { get }
           ^
error: terminated(1): /opt/swift/usr/bin/swift-build-tool -f /homer/leif/swift/Kitura/.build/debug.yaml
```

**Comment by backtrack94** [comment]
i quote this issue, i have too this problem with SSL on Kitura. I hope someone can get us an answer early.

**Comment by artemSydorenko** [comment]
I have same errors on Linux, OpenSSL 1.1.0f

**Comment by artemSydorenko** [comment]
I solved it: `sudo apt-get install libssl1.0-dev`

**Comment by zwoop** [comment]
Yeh, but that's a pretty weak solution. We should leave this open, for a real fix for building against OpenSSL v1.1.x. :-)

**Comment by billabt** [comment]
👍 I agree. Hopefully I have a chance to look at it soon.

**Comment by billabt** [comment]
As of now, there's no way to support _both_ the 1.0.x and 1.1.x versions of OpenSSL in the same module using the same interface. I'm investigating creating a branch of this framework that will support _OpenSSL v1.1.x_ that will be separate from the current version.  The current version, _1.0_, of _BlueSSLService_ will be used for OpenSSL 1.0.x while a new _2.0_ version of _BlueSSLService_ will be used to support OpenSSL 1.1.x.  No estimate at this time as to when this will available.  I hope to be able to start on it soon.  In the meantime, I've marked this issue as an **enhancement**.

**Comment by moshegottlieb** [comment]
> As of now, there's no way to support _both_ the 1.0.x and 1.1.x versions of OpenSSL in the same module using the same interface. I'm investigating creating a branch of this framework that will support _OpenSSL v1.1.x_ that will be separate from the current version. The current version, _1.0_, of _BlueSSLService_ will be used for OpenSSL 1.0.x while a new _2.0_ version of _BlueSSLService_ will be used to support OpenSSL 1.1.x. No estimate at this time as to when this will available. I hope to be able to start on it soon. In the meantime, I've marked this issue as an **enhancement**.

Another solution would be to create a proxy, a small C library that exposes the API need for BlueSSL and have it configured to use the available openssl version installed on the local machine (and maybe even common crypto on macOS instead of openssl).

**Comment by nethraravindran** [comment]
https://github.com/IBM-Swift/BlueSSLService/pull/61 adds support for _openssl 1.1.x_

**Comment by nethraravindran** [comment]
Resolved in https://github.com/IBM-Swift/BlueSSLService/pull/61

**Comment by djones6** [comment]
Addressed in BlueSSLService `1.0.24` and newer.
